kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StateDeserializer.java

```diff
@@ -16,28 +16,45 @@
 
 package dev.responsive.kafka.internal.utils;
 
+import java.util.Optional;
 import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.apache.kafka.streams.state.internals.ValueAndTimestampSerde;
 
 public class StateDeserializer<K, V> {
   private final String changelogTopic;
   private final Deserializer<K> keyDeserializer;
   private final Deserializer<V> valueDeserializer;
+  private final Optional<Deserializer<ValueAndTimestamp<V>>> timestampedValueDeserializer;
 
   public StateDeserializer(
+      final boolean isTimestamped,
       final String changelogTopic,
-      final Deserializer<K> keyDeserializer,
-      final Deserializer<V> valueDeserializer
+      final Serde<K> keySerde,
+      final Serde<V> valueSerde
   ) {
     this.changelogTopic = changelogTopic;
-    this.keyDeserializer = keyDeserializer;
-    this.valueDeserializer = valueDeserializer;
+    this.keyDeserializer = keySerde == null ? null : keySerde.deserializer();
+    this.valueDeserializer = valueSerde == null ? null : valueSerde.deserializer();
+
+    if (isTimestamped && valueSerde != null) {
+      timestampedValueDeserializer =
+          Optional.of(new ValueAndTimestampSerde<>(valueSerde).deserializer());
+    } else {
+      timestampedValueDeserializer = Optional.empty();
+    }
   }
 
   public K keyFrom(final byte[] keyBytes) {
     return keyDeserializer.deserialize(changelogTopic, keyBytes);
```

Comment:
should we throw a custom exception if `keyDeserializer` is null instead of an NPE? (same for below)